include/random_walks/shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,199 @@
+
+// Contributed and/or modified by Iva Janković, as part of Google Summer of Code 2025 program.
+
+// Licensed under GNU LGPL-3.0; see LICENCE file
+
+#ifndef RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+#define RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+
+#include <Eigen/Eigen>
+#include <cmath>
+#include <algorithm>
+#include <stdexcept> 
+
+#include "sampling/sphere.hpp"
+#include "convex_bodies/hpolytope.h"
+#include "convex_bodies/correlation_matrices/corre_matrix.hpp"
+
+struct ShakeAndBakeWalk
+{
+    //enum Mode { Original, Limping, Running };
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point = typename Polytope::PointType;
+        using VT = typename Polytope::VT;
+        using NT = typename Point::FT;
+
+        struct update_parameters {
+            int   facet_prev   = -1;  
+        };
+
+        update_parameters params_;
+
+        static constexpr NT kDefaultEpsilon = NT(1e-10);
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope&          P,
+            const Point&             boundary_pt, 
+            int                      facet_idx,    
+            RandomNumberGenerator&   rng,
+            //Mode                     m   = Mode::Original,
+            NT                       eps = kDefaultEpsilon)
+            : P_{P}, /*mode_{m}*/ epsilon_{eps}
+        {
+            initialize(boundary_pt, facet_idx, rng);
+        }
+
+        void set_epsilon(NT eps) noexcept { epsilon_ = eps; }
+        NT   get_epsilon() const noexcept { return epsilon_; }
+
+         void apply(unsigned int walk_len, RandomNumberGenerator& rng)
+        {
+            const NT eps = epsilon_; 
+
+            for (unsigned step = 0; step < walk_len; ++step)
+            {
+                Point v = get_direction(rng);
+
+                int facet_new;
+                std::tie(lambda_hit_, facet_new) = P_.line_positive_intersect_skip(p_, v, Ar_, Av_, lambda_hit_, params_);
+
+                if (!std::isfinite(lambda_hit_) || lambda_hit_ <= eps  || facet_new < 0) 
+                {
+                    lambda_hit_ = NT(0);
+                    continue;
+                }
+
+                p_ += lambda_hit_ * v;
+                facet_idx_ = facet_new;
+                A_row_k_   = P_.get_facet_normal_vec(facet_idx_);
+                params_.facet_prev  = facet_idx_;
+            }
+        }
+
+
+        const Point& getCurrentPoint() const noexcept { return p_; }
+
+    private:
+
+        Point get_direction(RandomNumberGenerator& rng) 
+        {
+
+            VT ck = A_row_k_;
+            ck.normalize();
+
+            std::vector<NT> u(dim_);
+            for (unsigned int i = 0; i < dim_; ++i) {
+                u[i] = rng.sample_ndist();
+            }
+
+            NT dot = NT(0);
+            for (unsigned int i = 0; i < dim_; ++i) dot += u[i] * ck[i];
+            for (unsigned int i = 0; i < dim_; ++i) u[i] -= dot * ck[i];
```

Comment:
Plz use our Point class for these computations. For example `NT dot = u.getCoefficients().dot(A_row_k_);`